In [1]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn import svm
import numpy as np
import pandas as pd

### Loading dataset

In [2]:
diab_dataset = pd.read_csv('dataset/diabetes.csv')  # diabetes dataset

In [3]:
diab_dataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
diab_dataset.shape

(768, 9)

### Statistical measures of data

In [5]:
diab_dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [6]:
diab_dataset['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

## Outcome
0 rep. non-diabetic. In this case 500 people are non-diabetic
1 rep. diabetic. In this case 268 people are diabetic

#### Mean

In [7]:
diab_dataset.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

#### Separate data and labels

In [8]:
X = diab_dataset.drop(columns='Outcome', axis=1)  # if you're dropping a column, axis=1, if you're dropping a row axis=0
Y = diab_dataset['Outcome']

In [9]:
print(X)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [10]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


## Data standardization
The range is between all values is too wide. Standardize the range for accurate predictions.

In [11]:
scaler = StandardScaler()

In [12]:
scaler.fit(X)

StandardScaler()

In [13]:
std_data = scaler.transform(X)   # standardized data

In [14]:
print(std_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


#### Data printed above are in the range 0 - 1.

In [15]:
X = std_data
Y = diab_dataset['Outcome']

### Split data into training data and test data

In [16]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)
# test_size = 20% of the dataset is used to train the model
# stratify all diabetes cases may be stored in X_train while non-diabetic will be stored in Y_train
# random_state ensures that a dataset will be splitted differently
print(f'X_train: {X_train}')
print(f'X_test: {X_test}')
print(f'Y_train: {Y_train}')
print(f'Y_test: {Y_test}')

X_train: [[-1.14185152 -0.05929342 -3.57259724 ...  0.05170968 -0.9992857
  -0.78628618]
 [ 0.63994726 -0.49745345  0.04624525 ... -0.15136112 -1.05666795
   0.31985461]
 [-0.84488505  2.13150675 -0.47073225 ... -0.24020459 -0.2231152
   2.19178518]
 ...
 [ 2.12477957 -1.12339636  0.25303625 ... -0.24020459 -0.51908683
   0.14967911]
 [ 0.04601433 -0.27837344  0.45982725 ...  0.94014439 -0.71237443
   0.40494237]
 [-1.14185152 -1.09209922 -0.05715025 ...  0.48323511 -0.70633419
  -0.70119842]]
X_test: [[-0.25095213 -0.46615631  0.14964075 ... -0.78595734 -0.79995787
  -0.53102292]
 [-0.25095213 -0.24707629 -1.29789624 ... -1.21748278 -1.00230582
  -0.95646168]
 [ 0.3429808   0.81702665  0.45982725 ...  0.2167047  -0.76673656
   2.7023117 ]
 ...
 [-0.54791859 -0.34096773  0.25303625 ...  0.05170968  0.6829204
  -0.53102292]
 [-0.54791859 -1.24858494  0.14964075 ... -0.24020459  0.22688248
  -0.70119842]
 [ 2.7187125  -0.46615631  0.14964075 ...  0.58477051 -0.88754131
   1.00055664]]
Y_

In [17]:
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


### Training the model

In [18]:
classifier = svm.SVC(kernel='linear')  # SVC - Support Vector Classifier

### Training support vector machine classifier

In [19]:
classifier.fit(X_train, Y_train)

SVC(kernel='linear')

### Model evaluation
#### Accuracy score

In [20]:
# accuracy score - training data
X_train_prediction = classifier.predict(X_train)
training_data_acc = accuracy_score(X_train_prediction, Y_train)  # training data accuracy

In [21]:
print(f'Accuracy score: {training_data_acc}')

Accuracy score: 0.7866449511400652


#### Accuracy score - test data

In [22]:
X_test_prediction = classifier.predict(X_test)
test_data_acc = accuracy_score(X_test_prediction, Y_test)
print(f'Accuracy score (test data): {test_data_acc}')

Accuracy score (test data): 0.7727272727272727


## Making a predictive system

In [23]:
input_data = (3,158,70,30,328,35.5,0.344,35)

# convert input_data to numpy array
input_data_np = np.asarray(input_data)

# reshape array to make prediction for one instance
reshaped_input_data = input_data_np.reshape(1, -1)

# standardize input data
_std_data = scaler.transform(reshaped_input_data)
print(_std_data)
# Make prediction
_prediction = classifier.predict(_std_data)
print(_prediction)

if _prediction[0] == 0:
    print('Non-diabetic')
else:
    print('Diabetic')


[[-0.25095213  1.16129525  0.04624525  0.59362962  2.15509929  0.44515934
  -0.38620161  0.14967911]]
[1]
Diabetic


/home/brocode/Documents/Coding/Python/ml-projects/.venv/lib/python3.11/site-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
